app/src/main/java/org/thoughtcrime/securesms/util/FlowUtils.kt

```diff
@@ -0,0 +1,42 @@
+package org.thoughtcrime.securesms.util
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.flatMapConcat
+
+/**
+ * Buffers items from the flow and emits them in batches. The batch will have size [maxItems] and
+ * time [timeoutMillis] limit.
+ */
+fun <T> Flow<T>.timedBuffer(timeoutMillis: Long, maxItems: Int): Flow<List<T>> {
+    return channelFlow {
+        val buffer = mutableListOf<T>()
+        var bufferBeganAt = -1L
+
+        collectLatest { value ->
+            if (bufferBeganAt < 0) {
+                bufferBeganAt = System.currentTimeMillis()
+            }
+
+            buffer.add(value)
+
+            if (buffer.size < maxItems) {
+                // If the buffer is not full, wait until the time limit is reached
+                delay((System.currentTimeMillis() + timeoutMillis - bufferBeganAt).coerceAtLeast(0L))
```

Comment:
```suggestion
                delay(System.currentTimeMillis() + timeoutMillis - bufferBeganAt))
```

Don't need it to be non-negative, `<= 0` returns immediately.

Reply:
I did look at the source, haha, yeah, but it's also documented:

delay
suspend fun [delay](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/delay.html)(timeMillis: [Long](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-long/index.html))([source](https://github.com/kotlin/kotlinx.coroutines/tree/master/kotlinx-coroutines-core/common/src/Delay.kt#L121))
Delays coroutine for at least the given time without blocking a thread and resumes it after a specified time. If the given [timeMillis](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/delay.html) is non-positive, this function returns immediately.